### DISCUSSION TITLE
Speed up proof & price addition

### DISCUSSION THREAD

**Discussion by raphodn** [openfoodfacts/open-prices#526] [General]
### Currently

@serpico (extracted from https://github.com/openfoodfacts/open-prices/discussions/523)

> In the same spirit/related, I was adding prices yesterday based on price tag's pictures I took the same day, for reference I timed myself doing it, for 16 prices it took me 54'( with a some back and forth between OFF and Open Price webapp to create products which didn't exist, but I believe that's not the bulk of the time spent ) so it took 3'40" per price, although I'm willing to admit my workflow could be enhanced, I do use a text file from which I copy/paste the EAN for the products I track, saving some time I think.
I noticed that for a session like that, the type of proof, the location, the date and currency would be the same and the user would save time if those were once entered at the beginning of the so-called "session" and "locked", some kind of streamlined interface without all those field after initially being entered, ideally making use of the keyboard's Tab key to jump from one field to the next...maybe that's mostly what we have for the receipt mode. An additional "lockable" parameter would be EAN as opposed to category - ideally jumping straight the the EAN input box to paste the relevant EAN.
After checking what the receipt mode UI looks like it would be a variant of it, if it's deemed worth exploring I'll expand in a specific issue and share screenshot of my vision of the UI, based on the current receipt mode.
Now the question is how many user engage ( percentage of prices coming from this opposed to receipt ) in this, how much time do we save ( here's I'm thinking a stopwatch feature could be useful although it might scare away user when they realize how much time they spend for so few prices - so maybe make it bonus interface ) and is it worth the time coding this specific interface.

### Related issues

- https://github.com/openfoodfacts/open-prices-frontend/issues/961
- https://github.com/openfoodfacts/open-prices-frontend/issues/983
- https://github.com/openfoodfacts/open-prices/issues/387#issuecomment-2386059566
- https://github.com/openfoodfacts/open-prices/discussions/523

**Comment by raphodn** [comment]
I would definitely be curious of a screenshot (or quick video ?) of your setup !

I presume you have 1 proof image per price in your case ? So it does require a bit of time to first upload the proof, and then add the single price. The current price addition workflow is indeed better for multiple prices.

This discussion would help in offloading the time spent per proof/price and leave it to the community. 3'40 per price is too long :cry: 

And we could use some image detection/cropping
- split images with multiple price tags, into images with only 1 single price tag, to make it more distributable on Hunger
- having OCR - for instance https://github.com/openfoodfacts/open-prices/issues/320 - would pre-fill the price & barcode forms, and have the users only "double check"

**Reply by raphodn** [reply]
fyi my current workflow : 
- i take pictures with my phone (stored in the gallery), and add the proofs in the OP web app via the phone. my pictures always have at least 2 or 3 price tags on them
- once home, I open my laptop, connect to my dashboard, and for each proof add the related prices
  - my laptop screen is small, so sometimes by zooming on my phone image (phone next to my laptop)
  - sometimes opening the proof image in full size and displaying it side by side (but the web app compresses the images, the quality is not always enough...)

**Reply by raphodn** [reply]
@serpico

> - I take pictures with my phone (stored in the gallery), for specific reasons I would rarely have 2 or 3 price tags in one picture and if I did I would only exploit the one I'm interested about, as you guessed it's almost always 1 price / 1 picture.
> - Once home, I open my laptop ( 12-inch screen, so who's got the smallest one ? 🤣 ) , copy the pictures from the gallery to a specific folder, open the first picture with an image viewer ( it loads all the pictures in that folder, so I can easily navigate the pictures of the day )
> - I also open a .txt file where I have most of my tracked products ( 1 line containing a short description and the EAN separated by several tabulation - ballpark 200 lines for products only interspersed with notes - somehow match my count of 183 displayed in my dashboard ) , ctrl+f of either the tail of the EAN or some character chain I know will give me the product I'm looking for so I can copy/paste EAN straight into the appropriate box when needed.

> I'll try your method ( phone to upload pic + laptop to add the price(s) ) next week and time myself, see if it provides a sizeable improvement  and report back.

> PS: I have prepared a longer version ( also for myself ) detailing the specific reasons for 1 pic / 1 price, associated with a sample of a few dozens of pictures illustrating why a 1 pic / several prices, I think wouldn't work for me, I'm willing to send it to you on your email, if you want us to dig deeper.

**Reply by raphodn** [reply]
don't have much time to answer right now, but I'd be willing to dig deeper.

In a single shop & day session, how many pictures do you take ?

If there existed a multi-upload proof page (similar to https://prices.openfoodfacts.org/proofs/add/single - yeah needs to be fixed - but allowing to select multiple images for the same shop & date) would it help you ? though afterwards you would need to go to your dashboard > proofs > for each one add the corresponding price ^^ so only a marginal time gain.. ?

**Reply by raphodn** [reply]
@serpico

> No worries, I took notes so we can postpone for later, whenever...

> I'd say 15 to 40 pictures, not all will be uploaded ( doubled for quality )...but I'd say 10-12 prices per session minimum and as many pictures.

> > If there existed a multi-upload proof page (similar to https://prices.openfoodfacts.org/proofs/add/single - yeah needs to be fixed - but allowing to select multiple images for the same shop & date) would it help you ? though afterwards you would need to go to your dashboard > proofs > for each one add the corresponding price ^^ so only a marginal time gain.. ?

> I think it'd help, but let's table that, first I need to try your workflow, see how it goes. 

> I'd really like to give you an embryo of a mock up but I need to really to decompose the process ( i.e : upload picture > copy  the EAN from .txt > hit tab  on OP > paste the EAN > ) I'll open another issue for me to work on that, if you could take a look at it whenever you can and tell me what you think.

> Here is the issue in question [961](https://github.com/openfoodfacts/open-prices-frontend/issues/961)

**Comment by serpico** [comment]
Here's some feedback after a few price logging sessions, using some of your advice regarding the workflow :

**A/ First session**

_1/_ Upload the proofs using the webapp on the phone ( same type, same folder for source image, same shop location, same date, same currency ) for all the 29 pictures, it took 20' so on average 41" per picture, including 13" really uploading ( when you hit the green button after selecting the file/image, the button is spinning and when it's over the "done" button is unlocked and you can go into the menu and start the next proof.

_2/_ Log The prices 
I followed your workflow, once all the proofs are uploaded I go into the proof gallery, bottom menu of one proof card > Add a price 

Hit "TYPE A BARCODE"
copy/paste the barcode
Hit enter
Click on the price input box ( to focus )
Type the price ( display the last known price for this product and this location and https://github.com/openfoodfacts/open-prices-frontend/issues/567 could be useful )
Hit upload
Hit done

Here there's probably something to do in term of keyboard shortcuts, where the focus gonna be/should be every time we validate each step, there's probably a few keystroke, pointer movement to be saved...

19 prices in 18'42" or 1122" 59" per price

**So combined, per proof+price, it averages 1'40" per price  ( upload and logging ).** 

This is, what I consider, a significantly favorable sample as I trimmed my .txt file containing the EANs so I only have products from this shop and sorted in the order I go through the shop - there's a caveat here, the order in which I upload the proof might have an slight impact, it'd better to have the same order on the .txt file and on the proof gallery. I also keep the new products ( the ones that are not on my EAN trimmed list/.txt file ) as process separately - and it takes a lot more time ( change the quantity on OFF, add the brand, create the product from scratch...), but it's significant in my case, I'd say for every 15 to 20 known product ( allow fast mode, copy paste EAN and that's it ) there are 3 to 5 which require the slow way, I have a ballpark somewhere, I'll try to look for it

**B/ Second session**

_1/_ I copied the proof pictures from the phone to the laptop in a dedicated folder, then upload the proofs using the browser of the laptop instead of the doing on the phone like in A/.
There's a noticeable improvement to upload x16 proofs it took 8'44" or 524" or 33"per upload ( this is including the step to create the folder in the laptop, copy from phone to laptop )

_2/_ log the prices : x13 prices in 12'17" soit 737" ou 57" per price

**Combined 1'30" per proof+price ( upload and logging )**

**C/ Third session** 

_1/_ Same method as in B/ ( copy pictures from phone to laptop, then upload from laptop ) 
upload 17 proofs ( some duplicates ) in 8'20" or 500" averages to 29" per proof

_2/_ log the prices : 11'40" for 11 prices or 700" averages 64" per price

**Combined 1'33" per proof+price ( upload and logging )**

One thing that still slows the process, is once the price is sent the dashboard page is automatically loaded, then you need to click the Proof button at the end of each price, load the gallery, look for the next proof - if we went back directly to the proof gallery we could save some time - even more if there was an option to display only the proofs with 0 price (having a "stack/deck" of proofs to go through for one session, the stack would slim down every time we log price - only proofs with 0 price would stay in the "stack/deck" to be exploited - so if you mess up the upload order compare to your list of EANs and proceed in reverse order, each time you need to load 1 or more block of X proofs - a more desirable situation would be the first ( at the top ) proof you load/see in the proof gallery is the one associated with the next EAN on your list. 

Hope it's not too messy.

**Comment by TTalex** [comment]
Related: I've noticed losing a bit of time when mistyping/misreading barcodes, so I've proposed a solution here: https://github.com/openfoodfacts/open-prices-frontend/pull/991

Note that this doesn't really help on the above feedback. It might even worsen it, since it breaks some of the enter key shortcuts.

**Reply by raphodn** [reply]
Can you explain a bit more ? You mistype/misread the barcode because the image is not very clear, or your hesitating between a 3 & a 6 or 8 for instance ? (edit: ok read your PR description, there's more details there :+1: )

I opened an issue in search-a-licious a few weeks ago to see what could be done for the fuzzy search, probably related :)
https://github.com/openfoodfacts/search-a-licious/issues/242
--> these complex product barcode queries could be sent to search-a-licious instead of OP backend (but we wouldn't have the price_count)

**Comment by raphodn** [comment]
With the recent "Contribution assistant" built by @TTalex , I'm seeing 2 (or 3) seperate steps that we could isolate, allowing distribution on multiple users.

Some of these thoughts have already been shared in a similar discussion - #523 - and some related issues.

## Step 1: upload 1 or multiple proofs [user]

Speed up proof addition when a user has multiple proofs for a given location & date.

### Related issues

- https://github.com/openfoodfacts/open-prices-frontend/issues/983

## Step 2: extract data from images [AI/ML]

- automatically predict the proof type
- automatically crop out price tags (bounding boxes)
- ask an AI to return the price predictions (see the current Contribution assistant that uses Gemini AI)

### Related issues

- https://github.com/openfoodfacts/open-prices/issues/394
- https://github.com/openfoodfacts/open-prices/issues/320
- https://github.com/openfoodfacts/open-prices-frontend/issues/1044
- #511 
- #611

## Step 3: ask users to validate the price predictions [user]

Last step, we have price info that has been extracted from the images, we need a user to validate and make sure that it is the correct product (or raw category) and price.

### Related issues

- https://github.com/openfoodfacts/open-prices/issues/568

**Reply by serpico** [reply]
Any chance to see some testing for #308 ( using the pictures already in OFF ( those containing a price sticker especially )) ? 

I'd really like to proof check/validate a limited sample of those pictures because there are prices that date back several years which we are unlikely to get otherwise ( sometimes I see for example fresh meat as cheap as it was several years ago on those pics but I can't firm up this impression easily unless more of those OFF's "old prices" get into open price somehow ).

I also really like @raphael0202's idea of different tier of quality ( "gold data" and so on ) mentioned [here](https://github.com/openfoodfacts/open-prices/issues/387#issuecomment-2385964339) in my mind those old prices coming from OFF's pictures could be in a lower tier and still be accepted since they don't always have the location ( sometimes though the shop's name and postcode is on the sticker - as for date of purchase could we use the timestamp/EXIF of the picture ? or some arbitrary date between the packaging date ( emballé le jj/mm/yy ) and the best before date ( à consommer avant le jj/mm/yy ) ).

But still it'd be great if they could show up on a graph so we get to see the trend.

**Reply by raphodn** [reply]
Quick answer on Hunger games, the issue is that it would require development on https://hunger.openfoodfacts.org, and I don't believe any one has much time at the moment...
So we opted to test things out in our "Experiment" sandbox - https://prices.openfoodfacts.org/experiments - at least for the time being !
We might end up with a https://hunger.prices.openfoodfacts.org :sweat_smile:

**Reply by serpico** [reply]
I'll raise it up again in a few months :)

**Reply by raphodn** [reply]
did a quick workflow using https://app.diagrams.net
raw file here if you want to edit/improve ! https://file.io/iUUHmOUxFfpk

![20241211_contribution_flow](https://github.com/user-attachments/assets/155cda41-7a3f-4d61-858f-6305364f1a67)

**Reply by raphodn** [reply]
Quick update : 

1) in the frontend the multiple proof upload page is live since a few days
- https://prices.openfoodfacts.org/proofs/add/multiple
- it's missing a check on the images (and some info message) to make sure they are all from the same date

2) in the backend we've unlocked 2 things
- proofs can now be uploaded anonymously
- any user can add prices on PRICE_TAG proofs

Next step: 
- ~frontend: a new contribution page displaying all PRICE_TAG proofs without any prices. And allowing any user to start adding prices~ (decided to move faster and go directly to the price-level instead of proofs)
- frontend: start splitting your price dashboard, between personal data (RECEIPT/GDPR proofs & their related prices), and common data (PRICE_TAG proofs or prices)
- backend: store price-level predictions
- frontend: a new contribution page displaying price tags to be validated (and effectively create prices from predictions)

**Reply by raphael0202** [reply]
Agree! I would also add as next step, add a multiple image upload form accessible without log-in for anonymous proof upload (as discussed last wednesday).

**Reply by raphodn** [reply]
fyi i updated the workflow, to show the intermediate "price tag" prediction

**Reply by raphodn** [reply]
Most of the work has been done in the backend during the last few days !!
https://github.com/openfoodfacts/open-prices/issues/611

**Reply by raphodn** [reply]
I just created an issue in the frontend to start working on a "price tag validation" UI
(restricted to only your proofs/prices for now)
https://github.com/openfoodfacts/open-prices-frontend/issues/1137

**Comment by serpico** [comment]
Should we split the discussion ? as :

- Speed up proof & price addition using AI
- Speed up proof & price addition by refining the UI [ full manual - one of the scenario I'm thinking about is the picture is so bad the person that took it can barely make out the EAN ( still can but also by trying to amend/correct the EAN and cross reference/check using brand, shop, whatever ) so what's AI gonna do...]

Example : In the above workflow I'd, personally, would put "Classify receipt/price tag" on the 1st step since when I have a batch of price/pictures to upload I do them all, without receipt in between...

**Reply by raphael0202** [reply]
Even if the EAN cannot be read, we can still provide product suggestion based on :

- the price tag label (ex : confiture mûre 400g Grosjean)
- the product that is associated with the price tag: we can find similar products in OFF based on the product image

We can't guarantee a 100% match rate, but at least we could provide valuable suggestions most of the time.
When ML fails, we should always let the user provide the data efficiently. ML-based contribution and manual contribution should be thought as a whole in my opinion.

**Comment by bobanmilano** [comment]
sorry, i'm not sure if my question is on the right place here. i tried for the first time to add a price but found out that i had to upload a "proof" image first. so i made the image (jpg, ~50KB) and tried to upload it via POST to https://prices.openfoodfacts.org/api/v1/proofs/upload. 

but i get a 500 server error each time. is the server not working correctly or am i doing something wrong here?

**Reply by raphodn** [reply]
Hi ! I had a look at the server errors and it seems the date field is missing. I'll try to take some time this weekend and improve the error mgmt/messages for this endpoint 👌

**Reply by bobanmilano** [reply]
thank you very much! indeed the date field was missing, now it had worked, i was able to upload the proof and the price! :) but it seems somehow not to be stable - now i get "connection reset by peer" errors 

(ClientException with SocketException: Connection reset by peer (OS Error: Connection reset by peer, errno = 104), address = prices.openfoodfacts.org)

what i am trying to do: i am creating a smartphone app for comparing prices in austria and germany as we have large price differences which no one can really explain. so the goal is to collectively put some presser (via social media, email, ....) on the chains and to let them now that we the customers are not satisfied! 
i want to provide the users with the ability to add prices and afterwards to share their findings on social media and/or write an email to the customer support of  the corresponding shop chain. may i ask what the best solution is for this purpose:

- having all the users post prices via my openprices account
- or each user create an own account on "openprices"

i want to make it as easy as possible for the user and doing the authentication twice (for the app and for openprices) would be confusing to the user i guess. additionally i will be storing each users priceaddition so that every added price can be tracked to the specific user (to prevent misusage).

**Reply by raphodn** [reply]
I pushed a fix in the [latest release](https://github.com/openfoodfacts/open-prices/pull/1060) (regarding the 500 error)

for the "connection errors" it might be some hiccups on the OFF server side :eyes:

**Reply by raphodn** [reply]
great to hear about your app idea!
we have a weekly meeting on Thursdays at 2pm, want to join? or discuss this in a seperate/quick meeting that day?

some quick answers in the meantime:
- you would have your own backend, or would solely rely on openprices DB ?
- for authentication: openprices account means openfoodfacts account, it's the same user base
- and usually apps that leverage (and contribute) openfoodfacts data don't indeed ask for a double authentication, and send data via one "app" account, with a user id as well to isolate potential issues and easily revert/delete data
- if prices added are rather for "communicating" then "tracking my consumption", i guess not having per-user price stats shouldn't bother too much

**Reply by raphodn** [reply]
Heads up:
- proofs are considered valid if they are photos of price tags in stores, or photos of receipts (or GDPR requests of fidelity cards) (or authorized import of shop). The first proof you uploaded of Nutella, without any shop info nor currency, and taken from the internet, will be flagged as invalid once we have a moderation mechanism (soon)
- we highly encourage having a location for each proof. Without, well the data will be inexploitable / of lower quality
- if you want to test the API, there is a pre-prod environment : https://prices.openfoodfacts.net
- are you on Slack ? There is a #prices channel
- if you have any questions or struggle or lack of docmentation, tell us, we haven't prioritized it much yet..

How did you get to know the project btw ?

**Reply by bobanmilano** [reply]
yeah, thanks for the invitation. i would love to join your meeting at 2pm on thursday!

@own backend: only something very basic, probably just firebase to handle useraccounts (login, registration ....)
@openfoodfacts: i know it's the same userbase, i once made a simple "product information app" using openfoodfact-API (only read, no write)
@price stats: we don't need that. it's really only about comparing prices in ATD and DE and post the result on social media, very simple. i am doing this in my free time, so no big time and ressources investment - just a simple app that can be made and launched within a couple of days.
@proofs: as soon as the app is in use there will be - hopefully - lot's of austrians living in the border region to germany who will be creating real price updates from real shops and making the photo of the product and price within the app
@documentation: thank you very much, i really appreciate your support. there are questions all the time :) like e.g. now somehow the display of the "proof" image is not working. i also tried to find the #prices channel on slack but it could not be found. do i need an invitation for that channel?

![Screenshot_20251012_104310_com example food_prices_app 1](https://github.com/user-attachments/assets/46a5a828-6de1-4aba-a59c-1b92a33b31f2)


i found "openprices" when i searched the internet for a way to fetch prices from germany. but there is no easy way to do that. the shop chains are technically doing everything that's possible to even prevent webscraping and they have no public database with prices. so i found on the internet the "openprices" project and it would solve perfectly the goal to compare prices (as soon as enough prices from germany and austria are available). 

the difference in prices between austria and germany is up to 100% and this fact is these days discussed very heavily in politics. even austrian products are cheaper in germany than in austria - can you believe that? 😅

so, i hope to make some change in this whole unfair situation by enabling the customers to put some presser on the shop chains by posting their prices on social media and flooding their email-accounts with customer protests. 😁

**Reply by raphael0202** [reply]
@bobanmilano The link of the #prices Slack channel: https://openfoodfacts.slack.com/archives/C02Q6KM7F

Regarding the proof image, can you display the URL of the image you try to display? And do you have a status code by any chance?

**Reply by bobanmilano** [reply]
i have decided - considering all the overhead and additional steps for the app-user when working with openprices - to use firebase as a way to store prices. the reason is: my app does not need a 100% proof where the price is found and so on. it's really only about recording price differences (which everyone can also find in german and austrian webshops and be able to record them in the app) and post them on social media.
the "time window" for such an app is very short as it is currently a very hot topic here up to the highest levels of politics. in 2-3 months the topic will be probably completely forgotten.
so the "time question" and the simplicity for the user are the main reasons for deciding do use a custom backend. but thanks anyway for your kind support. i think openprices and openfoodfacts are great projects and i wish you all the best! 
kind regards
boban

**Reply by raphael0202** [reply]
I understand, good luck with your project! If you happen to develop a more long-term solution for price tracking, we will always welcome contributions from external apps :)
